### PR TITLE
add last seen field to client (fixes #400)

### LINKED ISSUE #400: last message/last ping in /clients
**Is your feature request related to a problem? Please describe.**
If you use `gotify` to handle notifications it would be nice to know if clients are still reachable (or have the app installed, app killed). You can then fallback to another channel (email, sms ...) to not miss updates.

**Describe the solution you'd like**
`/client` already returns the list of clients, add another field like `last_message` or `last_keepalive`/`lastping` to show when that client has last received a message or last responded to a ping.

### PR DESCRIPTION
**Pull request by eternal-flame-AD**
This PR fixes #400 by adding a last seen field to the client model. The last seen field is updated on connection & every pong response.

### CODE REVIEW

File: model/client.go [hunk 1]
```diff
@@ -24,4 +26,9 @@ type Client struct {
 	// required: true
 	// example: Android Phone
 	Name string `gorm:"type:text" form:"name" query:"name" json:"name" binding:"required"`
+	// The last time the client was seen.
+	//
+	// read only: true
+	// example: 2019-01-01T00:00:00Z
+	LastSeen *time.Time `json:"lastSeen"`
```
Comment (jmattheis) [inline model/client.go:33]: I think I'd prefer the name LastUsed, then it's more related to the token.

File: model/client.go [hunk 2]
```diff
@@ -24,4 +26,9 @@ type Client struct {
 	// required: true
 	// example: Android Phone
 	Name string `gorm:"type:text" form:"name" query:"name" json:"name" binding:"required"`
+	// The last time the client was seen.
```
Comment (jmattheis) [inline model/client.go:29]: Can we also update last seen when the token was used for api calls. I think the update could be done in auth/authentication.go in the RequireClient call. Similar to what I described in this comment: https://github.com/gotify/server/issues/400#issuecomment-1247082718

File: ui/src/client/Clients.tsx [hunk 3]
```diff
@@ -119,6 +123,7 @@ const Row: SFC<IRowProps> = ({name, value, fEdit, fDelete}) => (
                 style={{display: 'flex', alignItems: 'center', width: 200}}
             />
         </TableCell>
+        <TableCell>{lastSeen ? <TimeAgo date={lastSeen} /> : 'Never'}</TableCell>
```
Comment (jmattheis) [inline ui/src/client/Clients.tsx:126]: Maybe we can do something like if (now - lastSeen) < 5min then it only displays the text "recently".

File: router/router_test.go [hunk 4]
```diff
@@ -332,12 +336,30 @@ func (s *IntegrationSuite) TestSendMessage() {
 	json.NewDecoder(res.Body).Decode(token)
 	assert.Equal(s.T(), "backup-server", token.Name)
 
+	req = s.newRequest("POST", "client", `{"name": "mock-client"}`)
```
Comment (jmattheis) [inline router/router_test.go:339]: This feels a little out of place here, the test only tests the message api and now it also checks clients. I think it'd be enough to check if last seen was updated in the authentication package, and that the proposed GetActiveClientTokens method works correctly one of the stream tests.

**Comment by eternal-flame-AD** [comment]
Thanks for the review, its been some time since I was active here. Yeah I should have read #400 more carefully you are correct we are discussing more about token last used instead of client alive-or-not.

File: router/router_test.go (hunk 4 above)
Comment (eternal-flame-AD) [reply router/router_test.go]: I agree, this was done to make coverage happy,

File: router/router_test.go [hunk 5]
```diff
@@ -400,6 +404,12 @@ func (s *IntegrationSuite) TestAuthentication() {
 	assert.Equal(s.T(), "android-client", token.Name)
 }
 
+func (s *IntegrationSuite) connectWebSocket(url string) *websocket.Conn {
```
Comment (jmattheis) [inline router/router_test.go:407]: Unused, the changes in this file can probably be reverted.

File: ui/src/client/Clients.tsx [hunk 6]
```diff
@@ -119,6 +123,17 @@ const Row: SFC<IRowProps> = ({name, value, fEdit, fDelete}) => (
                 style={{display: 'flex', alignItems: 'center', width: 200}}
             />
         </TableCell>
+        <TableCell>
```
Comment (jmattheis) [inline ui/src/client/Clients.tsx:126]: The token overlaps the last used column, we could you make this non overlapping?

image

File: ui/src/application/Applications.tsx [hunk 7]
```diff
@@ -172,6 +187,17 @@ const Row: SFC<IRowProps> = observer(
             <TableCell>
                 <CopyableSecret value={value} style={{display: 'flex', alignItems: 'center'}} />
             </TableCell>
+            <TableCell>
```
Comment (jmattheis) [inline ui/src/application/Applications.tsx:190]: I prefer having the last_used column in the column before the controls (delete/edit).

**Review by jmattheis — CHANGES_REQUESTED** [review]

**Review by jmattheis — DISMISSED** [review]
:+1: I've also extracted the ui code for the last used cell to a separate file.

**Pull request merged**